cpp/ql/src/semmle/code/cpp/rangeanalysis/Bound.qll

```diff
@@ -0,0 +1,75 @@
+import cpp
+private import semmle.code.cpp.ir.IR
+
+private newtype TBound =
+  TBoundZero() or
+  TBoundInstruction(Instruction i) {
+    i.getResultType() instanceof IntegralType or
+    i.getResultType() instanceof PointerType
+  } or
+  TBoundOperand(Operand o) {
+    o.getDefinitionInstruction().getResultType() instanceof IntegralType or
+    o.getDefinitionInstruction().getResultType() instanceof PointerType
+  }
```

Comment:
I had thought that bounding relative to `Operand` might give a tighter bound than `Instruction` in some cases involving guard conditions; I'm no longer convinced that that's true.

Reply:
It also looks like `TBoundInstruction` includes too much if it's trying to correspond to `TBoundSsa` - it should likely exclude temporary variables that are introduced by the IR construction.  Note, that the overall algorithm works by constructing a graph (that's all the `*step*` predicates), which should be fast as its size is likely linear in the size of the program, and the recursive calculation in the graph essentially calculates the transitive closure in the graph restricted to certain starting points, and the `Bound` class is exactly those starting points.  So `Bound` should be as small as possible to maintain good performance, while including all the interesting nodes in the graph that one might want as a bound for something else, and all phi nodes as they are needed as bounds for themselves to get any sort of relevant bounds in loops.

Reply:
I absolutely don't think value numbers in general should be part of `Bound` - that's going to be too big without any benefits.  E.g. bounding anything by the value number of `x+1` is never going to be useful, as it's always better to bound directly by `x`.

Reply:
Right, sorry, just now read the commit with the code change above to use value numbers.  When restricted to atomic things, then it looks like it might be fine.  I'm slightly concerned by the `i.getAUse() instanceof ArgumentOperand` disjunct - doesn't that open the door for valuenumbers of things like `x+1`?

Reply:
Yes; Consider the following code:
``
int *foo(int x) {
  int *arr = new int[x*2];
  for(int i = 0; i < x*2; i++) {
    arr[i] = i; // we need to be able to relate i to the size of arr, which is x+1
  }
  return arr;
}
```
I think we will need to be able to have bounds relative to arbitrary `ValueNumber`s that are used as arguments to allocating functions, and given that C++ allows custom allocators, I don't think we'll be able to do that without allowing any `ValueNumber` used as an argument to be a bound.

Reply:
Ok, I can see why it can be useful.  Then let me just say that the size of the `Bound` type and whether a smaller set of valuenumbers can be used here is something to potentially revisit if performance becomes a problem.

Reply:
I'm starting on performance testing; it looks like the main stage of range analysis is about 50% slower with the `i.getAUse() instanceof ArgumentOperand` disjunct than without.